tests/integration/test_tenants.py

```diff
@@ -1,89 +1,67 @@
-    empty_test_db,
-    auth_headers,
-    valid_header,
-    create_property,
-    create_join_staff,
-):
-    staff_1 = create_join_staff()
-    staff_2 = create_join_staff()
-    newTenant = {
-        "firstName": "Jake",
-        "lastName": "The Dog",
-        "phone": "111-111-1111",
-        "staffIDs": [staff_1.id, staff_2.id],
-    }
-
-    newTenantWithLease = {
-        "firstName": "Finn",
-        "lastName": "The Human",
-        "phone": "123-555-4321",
-        "propertyID": create_property().id,
-        "occupants": 3,
-        "dateTimeEnd": Time.one_year_from_now_iso(),
-        "dateTimeStart": Time.yesterday_iso(),
-        "unitNum": "413",
-    }
-
-    response = client.post(endpoint, json=newTenant, headers=valid_header)
-
-    assert is_valid(response, 201)  # CREATED
-    assert response.json["firstName"] == "Jake"
-
-    response = client.post(endpoint, json=newTenantWithLease, headers=valid_header)
-    assert is_valid(response, 201)
-    assert response.json["unitNum"] == "413"
-
-    response = client.post(endpoint, json=newTenant, headers=valid_header)
 
 
 @pytest.mark.usefixtures("client_class", "empty_test_db")
-class TestTenant:
+class TestGet:
     def setup(self):
         self.endpoint = "/api/tenants"
 
-    def test_update_tenant(self, empty_test_db, valid_header, create_tenant):
+    def test_get_tenant(self, valid_header, create_tenant):
         tenant = create_tenant()
-        updated_fields = {"archived": True}
 
-        with patch.object(TenantModel, "update", return_value=tenant) as mock_update:
-            response = self.client.put(
-                f"{self.endpoint}/{tenant.id}",
-                json=updated_fields,
-                headers=valid_header,
-            )
+        response = self.client.get(f"{self.endpoint}/{tenant.id}", headers=valid_header)
+
+        assert response.json == tenant.json()
+
+
+@pytest.mark.usefixtures("client_class", "empty_test_db")
+class TestPut:
+    def setup(self):
+        self.endpoint = "/api/tenants"
+
+    def test_update_tenant(self, valid_header, create_tenant):
+        tenant = create_tenant()
+        updated_fields = {"archived": True}
 
-        mock_update.assert_called_once_with(
-            schema=TenantSchema, id=tenant.id, payload=updated_fields
+        response = self.client.put(
+            f"{self.endpoint}/{tenant.id}",
+            json=updated_fields,
+            headers=valid_header,
         )
-        assert response.status_code == 200
+
         assert response.json == tenant.json()
+
+
+@pytest.mark.usefixtures("client_class", "empty_test_db")
+class TestGetAll:
+    def setup(self):
+        self.endpoint = "/api/tenants"
+
+    def test_get_all_tenants(self, valid_header, create_tenant):
+        create_tenant()
+        create_tenant()
+        all_tenants = TenantModel.query.json()
+
+        response = self.client.get(self.endpoint, headers=valid_header)
+
+        assert response.json["tenants"] == all_tenants
```

Comment:
So right here, you're not really testing anything. You've written the same exact logic that is in the /tenants/ GET endpoint that we're trying to test:

```python
TenantModel.query.json()
```

Because you've duplicated the logic from the resource in this test, the only thing this test is doing is essentially making sure that the Flask-Restful Resource works.

Can you think of a different way to check that the proper data is returned in the response, regardless of the implementation?

Reply:
Is this more of what you're looking for?

```python
    def test_get_all_tenants(self, valid_header, create_tenant):
        tenants = [create_tenant().json(), create_tenant().json()]

        response = self.client.get(self.endpoint, headers=valid_header)

        assert response.json["tenants"] == tenants
```

Reply:
I tried using the patch.object framework
```
        with patch.object(TenantModel, "query") as mock_query:
            response = self.client.get(self.endpoint, headers=valid_header)
```

but I kept getting errors:
```
TypeError: Object of type MagicMock is not JSON serializable
```

Reply:
> 
> 
> Is this more of what you're looking for?
> 
> ```python
>     def test_get_all_tenants(self, valid_header, create_tenant):
>         tenants = [create_tenant().json(), create_tenant().json()]
> 
>         response = self.client.get(self.endpoint, headers=valid_header)
> 
>         assert response.json["tenants"] == tenants
> ```

Yes, this is a much better test since it checks that the endpoint logic (`TenantModel.query.json()`) is behaving the way we expect it to.

In regards to the patching, I'm not sure that we should do so here.  
If we wanted to, we could mock the `NobiruQuery.json()` method here and assert that it was called once, or alternatively, we could mock the `TenantModel.json()` method and assert that the mock's `call_count` is 2, but I think both of those are more confusing than this test.